_posts/2020-06-21-ski-numerals.markdown

```diff
@@ -0,0 +1,923 @@
+```
+
+Looking more manageable now.
+We have a lot of argument reordering to play with too.
+Note that every time we create a new local variable, this is marked with "enter new scope", since what we'll actually do is put an inner singleton function which takes the arguments so far and the new variable and evaluates the later expression.
+
+</div>
+
+### Arithmetic helpers
+{: #arithmetic-helpers}
+
+{% include collapser.markdown %}
+
+<div>
+
+The arithmetic operations used here are successor, predecessor, addition, subtraction, multiplication, division, and exponentiation - the full set of basic operations, pretty much.
+I took the techniques from [the Wikipedia article on Church encoding](https://en.wikipedia.org/wiki/Church_encoding), though there may be other ways to implement these operations.
+
+| Operator | Definition | As SKI | Size |
+|--:|:-:|:-:|:--|
+| `[1]` | `I` | `I` | `1 = 1` |
+| `[T]` | `a,b -> a` | `K` | `1 = 1` |
+| `[F] = [0]` | `a,b -> b` | `KI` | `2 = 2` |
+| `[=0]` | `n -> n (K[F]) [T]` | `S(SI(K(K[F])))(K[T])` | `9 = 6 + #[F] + #[T]` |
+| `[>=]` | `n,m -> [=0]([-R]nm)` | `S(K(S(K[=0])))[-R]` | `64 = 4 + #[=0] + #[-R]` |
+| `[++]` | `n,f,x -> f(nfx)` | `S(S(KS)K)` | `5 = 5` |
+| `[+]` | `m,n -> m[++]n` | `SI(K[++])` | `8 = 3 + #[++]` |
+| `[*]` | `m,n,f,x -> m(nf)x` | `S(KS)K` | `4 = 4` |
+| `[--]` | `n,f,x -> n(g,h -> h(gf))(Kx)I` | `S(S(KS)(S(K(S(KS)))(S(S(KS)(S(K(S(KS)))(S(K(S(KK)))(S(S(KS)K)(K(S(K(S(K(SI))))(S(K(S(KK)))(S(K(SI))K))))))))(K(KK)))))(K(K(KI)))` | `52 = 52` |
+| `[zf]` | `p -> pair (p(KI)) ([++](p(KI)))` | `S(S(KS)(S(K(SI))(S(KK)(SI(K(KI))))))(S(KK)(S(K[++])(SI(K(KI)))))` | `31 = 26 + #[++]` |
+| `[zp]` | `pair (KI) (KI)` | `S(SI(K(KI)))(K(KI))` | `9 = 9` |
+| `[--]` | `n -> n [zf] [zp] K` | `S(S(SI(K[zf]))(K[zp]))(KK)` | `48 = 8 + #[zf] + #[zp]`
+| `[-R]` | `m,n -> m[--]n` | `SI(K[--])` | `51 = 3 + #[--]` |
+| `[dR]` | `r,m,n -> [>=]nm([++](rrm(m[--]n)))[0]` | `S(S(KS)(S(K(S(KS)))(S(K(S(S(KS)(S(K(S[>=]))K))))(S(K(S(K(S(K[++])))))(S(S(KS)(S(K(S(KS)))(S(K(S(KK)))(SII))))(K(SI(K[--]))))))))(K(K(K[0])))` | `168 = 49 + #[>=] + #[++] + #[--] + #[0]`
+| `[/]` | `n,m -> (r -> rrmn)[dR]` | `S(S(KS)(S(K(S(S(KS)(S(K(S(SII)))K))))(S(KK)K)))(K(K[dR]))` | `191 = 23 + #[dR]` |
+
+Exponentiation doesn't require any special operator so it's not listed here.
+We avoid using the actual addition and subtraction operator, since we can usually save a few combinators by writing \\\(m[--]n\\\).
+
+</div>
+
+### Building Goodstein sequences
+{: #building-goodstein}
+
+{% include collapser.markdown %}
+
+<div>
+
+At a time like this, I really wish I had an optimizing SKI compiler.
+Here I am doing my best to hand optimize it, ignoring numerous other branches in the tree of possibilities which could possibly lead to a more optimized version, just taking the smallest expression I can get for each part and then linking with its dependencies.
+
+I am by no means claiming the final expression I get is optimal
```

Comment:
A period is needed at the end of the sentence: "...I get is optimal."

Reply:
nice work spotting that, thank you